### PR TITLE
[API Compatibility] align the loss apis --part

### PR DESCRIPTION
**Pull request by Manfredss**
### PR Category
User Experience

### PR Types
New features

### Description
align the loss apis in `nn.layer` and  `nn.functional`

### 是否引起精度变化
否

### CODE REVIEW

File: test/legacy_test/test_api_compatibility_part3.py [hunk 1]
```diff
@@ -2086,6 +2086,445 @@ def _test_static_Compatibility(self):
                 np.testing.assert_allclose(fetches[0], fetches[i], rtol=1e-5)
 
 
+class TestL1LossAPI(unittest.TestCase):
```
Comment (zhwesky2010) [inline test/legacy_test/test_api_compatibility_part3.py:2519]: 测下这些用法：size_average、reduce

**Review by zhwesky2010 — COMMENTED** [review]
这些没对齐的关键是 https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/utils/decorator_utils.py#L801 

Pytorch支持size_average、reduce的用法，用PaConvert先测试看看。

File: python/paddle/utils/decorator_utils.py
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py]: 这些装饰器体系有些复杂了，尽可能把这些装饰器合起来，代码复用：
1. 别名装饰器通过配置来解决，每个API只能有一个装饰器
2. func和layer的装饰器能否合并？

File: python/paddle/utils/decorator_utils.py [hunk 2]
```diff
@@ -865,37 +989,87 @@ def legacy_reduction_special_decorator(
...
+                cls_name == 'KLDivLoss'
+                and len(use_args) > 0
+                and use_args[0] in {'mean', 'sum', 'none', 'batchmean'}
+            )
+            if not ambiguous:
+                args, size_avg_val, reduce_val = (
+                    _consume_legacy_positional_args(cls_name, args, kwargs)
                 )
-            ):
-                raise_deprecated_error(cls_name, reduce_val, size_avg_val)
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py:892]: 看一看怎么在之前的基础上进行最小改动：只在之前的基础上设置下suggested，将raise_deprecated_error换个名字。

**Comment by Manfredss** [comment]
Paddle 与 PyTorch 在 loss API 上的不对齐主要分 5 类：
- (1) 已废弃的 `size_average/reduce` 参数 PyTorch 仍接受并翻译为 `reduction=`，Paddle 之前直接报错；
- (2) PyTorch 的位置布局比 Paddle 多 `size_average/reduce` 两个位置，且后续参数顺序也不同（如 `CrossEntropyLoss` 在 `reduction` 之后还插了 `soft_label`/`axis`/`use_softmax`）；
- (3) 参数名差异（`target` vs `label`、`beta` vs `delta`、`eps` vs `epsilon`、`anchor` vs `input` 等）；
- (4) API名差异（`multilabel_*` vs `multi_label_*`）；
- (5) `smooth_l1_loss` 默认公式不同（PyTorch 是非 Huber，Paddle `is_huber=True` 是 Huber）。


想法是把第 (1)(2)(3) 类一次性折叠到一个机制里：在 `decorator_utils.py` 中维护一张 `PYTORCH_LOSS_LAYOUT` 表，记录每个 loss API 位置布局；`legacy_reduction_decorator` 在入口处检查 `size_average` 槽是否为 `bool/None`，若是则判定为 PyTorch 风格调用，按 PyTorch 名字把所有位置参数转成 kwargs，再 pop 掉 `size_average`/`reduce` 翻译成 `reduction=` 并 raise DeprecationWarning；否则视作 Paddle 内部调用（如 `layer.forward` 调 `functional`），原样透传不动。这样位置歧义、内部调用冲突、参数重命名（交给下游的 `param_one_alias`）三个问题在同一处统一解决

第 (4) 类靠 `paddle.nn.functional.__init__` 里 4 行模块级别名解决；第 (5) 类用一个 11 行的 `smooth_l1_beta_compat` 装饰器：检测到 `beta` 时同时把 `is_huber` 默认设成 False，保留 Paddle 原有 `delta` 调用方的 Huber 默认。配套地，把 `PoissonNLLLoss`、`TripletMarginLoss` 两个 layer 上 `@param_one_alias(["epsilon", "eps"])` 与 `@legacy_reduction_decorator` 的顺序对调（前者改到内层）， 这样位置参数到关键字参数转换出来的 `eps` 才能被 `param_one_alias` 接力改名为 `epsilon`

**Comment by zhwesky2010** [comment]
> Paddle 与 PyTorch 在 loss API 上的不对齐主要分 5 类：
> 
> * (1) 已废弃的 `size_average/reduce` 参数 PyTorch 仍接受并翻译为 `reduction=`，Paddle 之前直接报错；
> * (2) PyTorch 的位置布局比 Paddle 多 `size_average/reduce` 两个位置，且后续参数顺序也不同（如 `CrossEntropyLoss` 在 `reduction` 之后还插了 `soft_label`/`axis`/`use_softmax`）；
> * (3) 参数名差异（`target` vs `label`、`beta` vs `delta`、`eps` vs `epsilon`、`anchor` vs `input` 等）；
> * (4) API名差异（`multilabel_*` vs `multi_label_*`）；
> * (5) `smooth_l1_loss` 默认公式不同（PyTorch 是非 Huber，Paddle `is_huber=True` 是 Huber）。
> 
> 想法是把第 (1)(2)(3) 类一次性折叠到一个机制里：在 `decorator_utils.py` 中维护一张 `PYTORCH_LOSS_LAYOUT` 表，记录每个 loss API 位置布局；`legacy_reduction_decorator` 在入口处检查 `size_average` 槽是否为 `bool/None`，若是则判定为 PyTorch 风格调用，按 PyTorch 名字把所有位置参数转成 kwargs，再 pop 掉 `size_average`/`reduce` 翻译成 `reduction=` 并 raise DeprecationWarning；否则视作 Paddle 内部调用（如 `layer.forward` 调 `functional`），原样透传不动。这样位置歧义、内部调用冲突、参数重命名（交给下游的 `param_one_alias`）三个问题在同一处统一解决
> 
> 第 (4) 类靠 `paddle.nn.functional.__init__` 里 4 行模块级别名解决；第 (5) 类用一个 11 行的 `smooth_l1_beta_compat` 装饰器：检测到 `beta` 时同时把 `is_huber` 默认设成 False，保留 Paddle 原有 `delta` 调用方的 Huber 默认。配套地，把 `PoissonNLLLoss`、`TripletMarginLoss` 两个 layer 上 `@param_one_alias(["epsilon", "eps"])` 与 `@legacy_reduction_decorator` 的顺序对调（前者改到内层）， 这样位置参数到关键字参数转换出来的 `eps` 才能被 `param_one_alias` 接力改名为 `epsilon`

方案基本没问题，可以再简化下：
- 第1~3类，给每个loss API维护一个别名表，也记录到LEGACY_POS中。对于关键字用法，则pop size_average/reduce，设置reduction，并修改别名。对于位置用法，由于目前的别名都是对应到了相同位置，应该别名不用做任何处理，而size_average/reduce直接去除，后面的参数位置参数也能对上，所以别名只需要删除size_average/reduce应该就可以？试试有无反例
- 第4类，处理好multi_label_的差异，然后新增API别名
- 第5类，单独处理，单独一个装饰器，参考index_add_decortor，看能否通过关键字与位置类型来区分出两者不同签名

File: python/paddle/utils/decorator_utils.py [hunk 3]
```diff
@@ -760,14 +760,25 @@ def wrapper(*args: _InputT.args, **kwargs: _InputT.kwargs) -> _RetT:
     return decorator
 
 
+# PyTorch positional layout for each loss API. Each entry maps the API
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py]: 注释可以不用写太多（只写必要的注释），从代码可读性与简洁的角度来设计代码，参考

https://github.com/pytorch/pytorch/blob/main/CLAUDE.md#coding-style-guidelines

File: python/paddle/utils/decorator_utils.py [hunk 4]
```diff
@@ -806,93 +860,157 @@ def compute_legacy_reduction(reduce_val, size_average_val):
...
     """
-    Function decorator for __init__: intercept deprecated 'reduce' and 'size_average'.
+    guard = LEGACY_SPECIAL_GUARD.get(name)
+    if guard is None:
+        return False
+    idx, allowed = guard
+    return len(use_args) > idx and use_args[idx] in allowed
+
+
+def legacy_reduction_decorator(fn=None, *, is_method=True):
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py]: 这个实现看起来也挺复杂的，那还是用第一版实现吧：

不用维护_SA0_RD1这些信息了，每个loss API维护一个args_list+kwargs_change，直接按args_list匹配然后pop掉size_average/reduce，设置reduction，修改别名即可。

相当于paconvert的逻辑搬过来

File: python/paddle/nn/layer/loss.py [hunk 5]
```diff
@@ -1072,6 +1073,7 @@ class PoissonNLLLoss(Layer):
     """
 
     @legacy_reduction_decorator
+    @param_one_alias(["epsilon", "eps"])
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py]: 每个API只允许一个装饰器

**Review by zhwesky2010 — COMMENTED** [review]
减少下注释吧。风格参考：
https://github.com/pytorch/pytorch/blob/main/CLAUDE.md#coding-style-guidelines

代码自身即为注释

File: python/paddle/nn/functional/__init__.py [hunk 6]
```diff
@@ -119,7 +119,9 @@
     margin_ranking_loss,
     mse_loss,
     multi_label_margin_loss,
+    multi_label_margin_loss as multilabel_margin_loss,
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/__init__.py]: 按这个来写别名

https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/functional/__init__.py#L178

保持统一风格

Image

File: python/paddle/nn/layer/loss.py [hunk 7]
```diff
@@ -2123,7 +2192,19 @@ class TripletMarginLoss(Layer):
...
+        args_list=[
+            'margin',
+            'p',
+            'eps',
+            'swap',
+            'size_average',
+            'reduce',
+            'reduction',
+        ],
+        kwargs_change={'eps': 'epsilon'},
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py]: 改成命名 alias_mapping 吧，按这个格式来

Image

File: python/paddle/nn/layer/loss.py [hunk 8]
```diff
@@ -2123,7 +2192,19 @@ class TripletMarginLoss(Layer):
     reduction: _ReduceMode
     name: str | None
 
-    @legacy_reduction_decorator
+    @legacy_reduction_decorator(
+        args_list=[
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py]: 改成命名 overload_args_list 吧

File: python/paddle/nn/layer/loss.py [hunk 9]
```diff
@@ -2827,3 +2929,9 @@ def predict(self, input: Tensor) -> Tensor:
                 output, indices, paddle.argmax(log_prob, axis=1).cast('float32')
             )
             return result
+
+
+# PyTorch-style aliases mirroring the functional layer's
+# ``multilabel_margin_loss`` / ``multilabel_soft_margin_loss`` aliases.
+MultilabelMarginLoss = MultiLabelMarginLoss
```
Comment (zhwesky2010) [inline python/paddle/nn/layer/loss.py]: 这个直接在\__init\__.py里写 a=b 这样吧

File: python/paddle/utils/decorator_utils.py [hunk 10]
```diff
@@ -760,140 +760,115 @@ def wrapper(*args: _InputT.args, **kwargs: _InputT.kwargs) -> _RetT:
...
-    )
-
-
 def legacy_reduction_decorator(
-    init_func: Callable[_InputT, _RetT],
-) -> Callable[_InputT, _RetT]:
-    """
-    Function decorator for __init__: intercept deprecated 'reduce' and 'size_average'.
-    """
+    fn=None,
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py]: 这个不需要了，第二层函数decorator接受func

File: python/paddle/utils/decorator_utils.py [hunk 11]
```diff
@@ -760,140 +760,115 @@ def wrapper(*args: _InputT.args, **kwargs: _InputT.kwargs) -> _RetT:
...
+                    DeprecationWarning,
+                    stacklevel=3,
                 )
-            ):
-                raise_deprecated_error(cls_name, reduce_val, size_avg_val)
-        return init_func(*args, **kwargs)
 
-    wrapper.__signature__ = inspect.signature(init_func)
-    return wrapper
+            renamed = False
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py]: 这个是想判断 smooth_l1_loss，如果是torch用法则进行配置is_huber

但是存在问题是无法完全识别torch的用法：
1. 使用了size_average是torch用法
2. torch关键字用法
3. torch位置用法
第3种情况无法区分是torch还是paddle，因此这个smooth_l1_loss可能需要改成 `paddle.compat.*`的形式，参考SKILL中的方案5

这里无法实现兼容

File: python/paddle/utils/decorator_utils.py [hunk 12]
```diff
@@ -760,140 +760,115 @@ def wrapper(*args: _InputT.args, **kwargs: _InputT.kwargs) -> _RetT:
...
+            ):
+                for i, val in enumerate(use_args):
+                    if i < len(args_list):
+                        kwargs.setdefault(args_list[i], val)
+                use_args = []
+
+            sa = kwargs.pop('size_average', '')
+            rd = kwargs.pop('reduce', '')
+            if (sa != '' and sa is not None) or (rd != '' and rd is not None):
+                suggested = compute_legacy_reduction(
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py]: 这里的辅助函数compute_legacy_reduction就两行，可以不需要
Image

File: python/paddle/utils/decorator_utils.py [hunk 13]
```diff
@@ -760,140 +760,115 @@ def wrapper(*args: _InputT.args, **kwargs: _InputT.kwargs) -> _RetT:
...
+                sa_idx >= 0
+                and len(use_args) > sa_idx
+                and (type(use_args[sa_idx]) is bool or use_args[sa_idx] is None)
+            ):
+                for i, val in enumerate(use_args):
+                    if i < len(args_list):
+                        kwargs.setdefault(args_list[i], val)
+                use_args = []
+
+            sa = kwargs.pop('size_average', '')
```
Comment (zhwesky2010) [inline python/paddle/utils/decorator_utils.py]: 这样写是否就行，这里的空字符串如果没有用可以简洁点

```
kwargs.pop('size_average', None)
rd = kwargs.pop('reduce', None)
```

File: python/paddle/nn/functional/loss.py [hunk 14]
```diff
@@ -1138,6 +1159,16 @@ def hsigmoid_loss(
         return out
 
 
+@legacy_reduction_decorator(
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/loss.py:1162]: smooth_l1_loss用装饰器无法实现对齐，因为必须要将is_huber设置为False，且对于<=2个位置参数的情况下无法判断torch用法和paddle用法。

所以需要改成新增一个paddle.compt.nn.functional.smooth_l1_loss的方式来对齐。

smooth这两个API单独提一个PR吧，这个PR先解决通过装饰器能对齐的问题

**Comment by zhwesky2010** [comment]
这个先合入了，smooth的对齐问题在下个PR里去解决。

**Review by zhwesky2010 — APPROVED** [review]
